src/overlays/actors/ovl_Elf_Msg2/z_elf_msg2.c

```diff
@@ -21,23 +31,172 @@ const ActorInit Elf_Msg2_InitVars = {
+        }
+        this->actor.shape.rot.x = this->actor.shape.rot.y = this->actor.shape.rot.z = 0;
+    }
+}
+
+void ElfMsg2_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Init.s")
+s32 ElfMsg2_GetMessageId(ElfMsg2* this) {
+    return (this->actor.params & 0xFF) + 0x100;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Destroy.s")
+/**
+ * Runs while navi text is up. Handles the killing of the actor upon closing the text box, can also set a switch flag
+ * from params.
+ */
+void ElfMsg2_Read(ElfMsg2* this, GlobalContext* globalCtx) {
+    s32 switchFlag;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809AD968.s")
+    // If the text box is closing this frame:
+    if (func_8002F334(&this->actor, globalCtx)) {
+        if (this->actor.posRot.rot.z != 1) {
+            Actor_Kill(&this->actor);
+            switchFlag = (this->actor.params >> 8) & 0x3F;
+            if (switchFlag != 0x3F) {
+                Flags_SetSwitch(globalCtx, switchFlag);
+            }
+        } else {
+            ElfMsg2_SetupAction(this, ElfMsg2_CheckForRead);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809AD978.s")
+/**
+ * Runs while navi text is not up. If text box flag is set, this function unsets it and proceeds to ElfMsg2_Read.
+ */
+void ElfMsg2_CheckForRead(ElfMsg2* this, GlobalContext* globalCtx) {
+    if (func_8002F194(&this->actor, globalCtx)) {
+        ElfMsg2_SetupAction(this, ElfMsg2_Read);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809AD9F4.s")
+void ElfMsg2_TextInit(ElfMsg2* this, GlobalContext* globalCtx) {
+    if ((this->actor.posRot.rot.y >= 0x41) && (this->actor.posRot.rot.y < 0x81)) {
+        if (Flags_GetSwitch(globalCtx, (this->actor.posRot.rot.y - 0x41))) {
+            ElfMsg2_SetupAction(this, ElfMsg2_CheckForRead);
+            this->actor.flags |= 0x00040001;
+            this->actor.textId = ElfMsg2_GetMessageId(this);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809ADA28.s")
+void ElfMsg2_Update(Actor* thisx, GlobalContext* globalCtx) {
+    ElfMsg2* this = THIS;
+    if (!ElfMsg2_KillCheck(this, globalCtx)) {
+        this->actionFunc(this, globalCtx);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Update.s")
+/**
+ * If nREG(87) is nonzero, a small translucent cuboid (with alpha = nREG(87)) is drawn around the check spot.
+ */
+void ElfMsg2_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    GraphicsContext* gfxCtx;
+    Gfx* dispRefs[4];
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Draw.s")
+    gfxCtx = globalCtx->state.gfxCtx;
+    Graph_OpenDisps(dispRefs, globalCtx->state.gfxCtx, "../z_elf_msg2.c", 355);
+    if (nREG(87) == 0) {
+        return;
+    }
```

Comment:
even though this makes more sense for the disps thing im wondering if this actually looks better. want to hear other opinions. Im leaning toward everything in the if block like it was before but idk

Reply:
Definitely early return imo

Reply:
Definitely early return imo